Refactor: 페이징 관련 리팩토링 (팔로잉/팔로워 목록 조회, 틈 요청 조회)

yangjiae12 opened the pull request:
### 👀 관련 이슈

<!-- 관련 이슈를 적어주세요 -->
#89 

### ✨ 작업한 내용

<!-- 작업한 내용을 적어주세요 -->
#### 공통
- 기존 페이징 0-based 방식 → 1-based로 통일
#### 팔로잉/팔로워 목록 조회
- `List` → `Slice`로 변경하여 페이징 최적화
- `PagingResponseDto<T>`라는 공통 응답 DTO를 `global.dto` 패키지에 분리해서 정의
- `content`와 `hasNext`만 포함하도록 응답 구조를 정리
#### 틈 요청 조회
- Page 그대로 유지
- ApiResponse 반환 시 `TeumSuccessStatus._TEUM_RECEIVED_LIST_SUCCESS` 명시
### 🌀 PR Point

<!-- 코드리뷰가 필요한 부분이 있다면 적어주세요 -->

### 🍰 참고사항

<!-- 참고할 사항이 있다면 적어주세요 -->

### 📷 테스트 결과
> API 명세서에서 자세하게 확인할 수 있습니다
#### 팔로잉 목록 조회
```json
{
  "isSuccess": true,
  "code": "FRIEND2002",
  "message": "친구 목록 조회에 성공하였습니다.",
  "result": {
    "content": [
      {
        "userId": 101,
        "nickname": "가영",
        "profileImageUrl": null,
        "isFavorite": true
      },
      {
        "userId": 104,
        "nickname": "라영",
        "profileImageUrl": null,
        "isFavorite": true
      },
      {
        "userId": 106,
        "nickname": "바영",
        "profileImageUrl": null,
        "isFavorite": true
      },
      // 중략...
      {
        "userId": 107,
        "nickname": "사영",
        "profileImageUrl": null,
        "isFavorite": false
      }
    ],
    "hasNext": true
  }
}
```
#### 팔로워 목록 조회
```json
{
  "isSuccess": true,
  "code": "FRIEND2002",
  "message": "친구 목록 조회에 성공하였습니다.",
  "result": {
    "content": [
      {
        "userId": 118,
        "nickname": "지우",
        "profileImageUrl": null
      },
      {
        "userId": 125,
        "nickname": "지원",
        "profileImageUrl": null
      },
      // 중략...
    ],
    "hasNext": false
  }
}
```
#### 틈 요청 조회
```json
{
  "isSuccess": true,
  "code": "COMMON2000",
  "message": "성공입니다.",
  "result": {
    "content": [
      {
        "responseId": 1,
        "requestId": 1001,
        "title": "요청 1",
        "description": "설명 1",
        "graphicId": 1,
        "senderUser": {
          "userId": 201,
          "nickname": "요청자1",
          "profileImageUrl": null
        },
        "receiverCount": 1,
        "date": "2025-07-25",
        "timeSlot": {
          "start": "12:00",
          "end": "13:00"
        },
        "read": false
      },
      {
        "responseId": 2,
        "requestId": 1002,
        "title": "요청 2",
        "description": "설명 2",
        "graphicId": 1,
        "senderUser": {
          "userId": 202,
          "nickname": "요청자2",
          "profileImageUrl": null
        },
        "receiverCount": 1,
        "date": "2025-07-25",
        "timeSlot": {
          "start": "12:00",
          "end": "13:00"
        },
        "read": false
      },
      // 중략...
      {
        "responseId": 9,
        "requestId": 1009,
        "title": "요청 9",
        "description": "설명 9",
        "graphicId": 1,
        "senderUser": {
          "userId": 209,
          "nickname": "요청자9",
          "profileImageUrl": null
        },
        "receiverCount": 1,
        "date": "2025-07-26",
        "timeSlot": {
          "start": "10:00",
          "end": "11:00"
        },
        "read": true
      }
    ],
    "pageable": {
      "pageNumber": 0,
      "pageSize": 10,
      "sort": {
        "empty": false,
        "sorted": true,
        "unsorted": false
      },
      "offset": 0,
      "unpaged": false,
      "paged": true
    },
    "last": true,
    "totalElements": 8,
    "totalPages": 1,
    "size": 10,
    "number": 0,
    "sort": {
      "empty": false,
      "sorted": true,
      "unsorted": false
    },
    "first": true,
    "numberOfElements": 8,
    "empty": false
  }
}
```

yangjiae12 committed:
Refactor: 틈 요청 조회 API의 페이지 요청을 1부터 시작하도록 변경

- page 파라미터를 1-based 기준으로 처리
- PageRequest 생성 시 page - 1 보정 적용

yangjiae12 committed:
Refactor: 팔로워 조회 API에 Slice 적용 및 페이지 1부터 시작으로 보정

- 기존 stream + subList 기반 페이징/정렬 로직 제거
- Spring Data JPA의 Pageable + Sort 조합 사용
- follower.nickname 기준 오름차순 정렬
- 클라이언트 페이지 번호는 1부터 시작하도록 처리

yangjiae12 committed:
Refactor: 팔로잉 조회 API에 Slice 적용 및 페이지 1부터 시작으로 보정

- 기존 stream + subList 기반 페이징/정렬 로직 제거
- Spring Data JPA의 Pageable + Sort 조합 사용
- isFavorite 기준 내림차순 -> nickname 기준 오름차순 정렬
- 클라이언트 페이지 번호는 1부터 시작하도록 처리

yangjiae12 committed:
Docs: Swagger 문서 개선을 위해 팔로워 조회 API에 @Parameter 설명 및 example 추가

yangjiae12 committed:
Refactor: 틈 요청 조회 API 페이지 번호 보정 책임을 서비스로 이동

- 컨트롤러에서는 page, size 그대로 전달
- 서비스 내부에서 1-based → 0-based 보정 처리
- DB 정렬 기준(readAt ASC, createdAt DESC) 포함

hyeonda02 commented on src/main/java/umc/teumteum/server/domain/friend/service/FriendServiceImpl.java:
```diff
@@ -61,39 +62,45 @@ public FavoriteResponseDto updateFavorite(Long userId, Boolean isFavorite) {
-    public List<FollowingUserResponseDto> getFollowingsByUser(Long userId, int page, int size) {
+    @Transactional(readOnly = true)
+    public Slice<FollowingUserResponseDto> getFollowingsByUser(Long userId, int page, int size) {
         validateUserExists(userId);
```
컨트롤러 단에서 이미 인증된 User 객체가 넘어오기 때문에 validateUserExists()를 서비스 단에서 다시 호출하는 부분은 중복 검증으로 보입니다! 해당 검증은 생략해도 괜찮을 것 같습니당

yangjiae12 replied:
인증 구조가 없을 때 userId만 파라미터로 받은 다음 서비스 단에서 유효성 검사를 했던 흐름이었는데 지금은 말씀 주신대로 이미 인증된 유저 객체를 받게 되어서 제거했습니다! 감사합니다 👏

hyeonda02 commented on src/main/java/umc/teumteum/server/domain/friend/service/FriendServiceImpl.java:
```diff
@@ -61,39 +62,45 @@ public FavoriteResponseDto updateFavorite(Long userId, Boolean isFavorite) {
                 .collect(Collectors.toList());
 
-        return paginate(sortedList, page, size);
+        return new SliceImpl<>(dtoList, pageable, slice.hasNext());
```
현재 Slice 객체를 그대로 응답에 포함하고 있어 페이징과 관련된 모든 데이터가 노출되는 구조입니다. 만약 프론트가 실제 사용하는 필드가 따로 있다면, 그 필드만 담은 DTO로 감싸는 구조도 고려해보시면 좋을 것 같습니다! 이렇게 하면 불필요한 데이터 전송을 최소화 할 수 있을 것 같습니다 :)

그리고 Slice를 그대로 반환하고 있는데 노션 명세서의 JSON 응답에는 hasNext 값이 없는 것 같습니다.
아마 hasNext()는 메서드라서 직렬화 대상에서 제외된게 아닐까.. 라는 의심이 드는데요..
무한 스크롤 페이징의 경우 hasNext는 거의 필수 값이라 포함해주는 것이 좋습니다.

따라서 응답 데이터를 별도의 DTO로 감싸고 hasNext를 포함한 필수 필드만 전달하는 구조를 도입해주시면 좋을 것 같습니다! 🥺 🙌

yangjiae12 replied:
말씀해 주신 것 바탕으로 `PagingResponseDto<T>`라는 공통 응답 DTO를 `global.dto` 패키지에 분리해서 정의했고 현제는 `content`와 `hasNext`만 포함하도록 응답 구조를 정리했습니다!

해당 구조로 테스트 진행하고 응답값은 PR에 반영해두었습니다 확인 부탁드립니다~ 

꼼꼼하게 봐주셔서 너무너무 감사해요 🥺

hyeonda02 reviewed (commented):
전에 말씀 드렸던 부분 리팩토링 해주셔서 감사합니다!
전체적으로 로직이 명확하게 잘 작성된 것 같습니당 수고 많으셨습니다 ~ 👍

yangjiae12 committed:
Refactor: 팔로잉/팔로워 목록 응답을 PagingResponseDto로 감싸도록 변경

- 기존 Slice<T> 응답 구조에서 PagingResponseDto<T>로 변경하여 불필요한 데이터 제거
- hasNext 필드를 명시적으로 포함해 무한스크롤에 필요한 최소 정보만 제공

hyeonda02 reviewed (approved):
수고 많으셨습니다! 🙌
